Feature/cli randomizer

sjb9774 opened the pull request:
Relating to this issue https://github.com/Ajarmar/universal-pokemon-randomizer-zx/issues/141

I'm not a Java developer by any means, so it's possible this will need some cleaning up. Additionally I'm not 100% sure if the code added for custom names is sufficient, it was borrowed from another place in the code but I can also see it's constructed a couple different ways and it wasn't immediately clear which was correct for this context. If it doesn't look right, I can look into it more.

I've tested this primarily with what I had on hand and am most familiar with, which is FireRed and it works nicely, but if there is a documented set of tests that should be run for contributions just let me know.

Also note, there seems to be an issue with the latest version of master loading the settings files I have on hand, so I've avoided merging that into this branch yet -- it seems independent of the changes made here.

Usage: `java -jar PokeRandoZX.jar cli <path to settings file> <path to source ROM> <path for new ROM>`

Ajarmar reviewed (changes requested):
I've had a look and also tested using this. This is a great start, but there are a few things that need to be fixed. If you have any questions about any of my comments, feel free to ask.

In the code-specific comments below, one of the things I mention would require the addition of another input parameter (specifically, saving 3DS games as a directory). With this in mind I also think it would be great if the input parameter structure could be reworked to use flags instead of positional arguments. This would make the CLI easier to use and also make it easier to add more input parameters if necessary. I'm thinking something like this in `invoke`:

```
        List<String> allowedFlags = Arrays.asList("-i", "-o", "-s", "-d");
        for (int i = 0; i < args.length; i++) {
            if (allowedFlags.contains(args[i])) {
                switch(args[i]) {
                    case "-i":
                        ...
                    case "-o":
                        ...
                    ...
                }
            }
        }
```

The flags I would suggest are:
- `-i` for input ROM
- `-o` for output ROM
- `-s` for settings file
- `-d` for saving as directory

So using the CLI randomizer would look something like this instead: 
`java -jar PokeRandoZX.jar cli -s <path to settings file> -i <path to source ROM> -o <path for new ROM> [-d]`

sjb9774 commented:
@Ajarmar Thanks for the feedback! I believe I've addressed it all and it should be ready for another review. I was close to re-inventing the wheel when it comes to parsing command line arguments but decided to just go with the simple route you suggested. If allowed, I'd be interested in exploring popular Java libraries (I assume this must exist) that handle it more elegantly.

sjb9774 commented on src/cli/CliRandomizer.java:
```diff
@@ -0,0 +1,223 @@
+            }
+
+            romHandler.loadRom(sourceRomFilePath);
+            CliRandomizer.displaySettingsWarnings(settings, romHandler);
```
@Ajarmar I don't love how some of these helper methods take the `RomHandler` directly and some take a ROM filepath and instantiate the ROMHandler from it but that's kind of required without the filepath/handler exposed from the ROMHandler itself. If there's no problem creating a public method for `RomHandler` that will return the file path then we could make these method signatures more consistent. It would be nice to be able to pass just the `RomHandler` since some methods may require `loadRom` to have been called and that can get very repetitive if we're doing that in every helper method.

Ajarmar replied:
To be honest - I don't think the helper methods help much here, I understand the desire to split things up into separate methods but the repeated creations of `File` and `RomHandler` objects just make the code more confusing to read. Your previous commit looked mostly fine; if you can go back to that one and push it, I have a few other things I'd like to give feedback on as well.

sjb9774 replied:
@Ajarmar Fair enough, most recent commit reverted

Ajarmar commented on src/cli/CliRandomizer.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    private static String usage() {
+        return "Usage: java -jar PokeRandoZX.jar cli -s <path to settings file> -i <path to source ROM> -o <path for new ROM> [-d save as directory]";
```
I'd rephrase this just to make it clear that the `-d` flag should be on its own, without anything extra after it. Also, to make it a bit simpler, `usage()` can just print directly instead of returning the string.
```suggestion
        System.err.println("Usage: java -jar PokeRandoZX.jar cli -s <path to settings file> -i <path to source ROM> -o <path for new ROM> [-d]");
        System.err.println("-d: Save 3DS game as directory (LayeredFS)");
```

Ajarmar commented on src/cli/CliRandomizer.java:
```diff
@@ -0,0 +1,194 @@
+
+    private final static ResourceBundle bundle = java.util.ResourceBundle.getBundle("com/dabomstew/pkrandom/newgui/Bundle");
+
+    private static boolean performDirectRandomization(
```
This method isn't used.

Ajarmar reviewed (changes requested):
Thanks for the changes, just a few more things I found that should be fixed. Also, there's one other thing that I thought about before but forgot to mention: Since the CLI randomizer now is part of the regular randomizer instead of being its own JAR, I think it shouldn't be in its own package but rather in the `com.dabomstew.pkrandom` package (so, specifically, it would be in `com.dabomstew.pkrandom.cli`).

Once again, if you have any questions, feel free to ask.

Edit: Oh, and also - I did look around a bit to see if there was a good way of handling CLI options in Java, but I'm reluctant to add external libraries and I don't think we really need anything more advanced than this implementation.

sjb9774 commented on src/cli/CliRandomizer.java:
```diff
@@ -0,0 +1,194 @@
+
+                    List<String> extensions = new ArrayList<>(Arrays.asList("sgb", "gbc", "gba", "nds", "cxi"));
+                    extensions.remove(romHandler.getDefaultExtension());
+                    File fh = FileFunctions.fixFilename(romFileHandler, romHandler.getDefaultExtension(), extensions);
```
@Ajarmar That makes a _lot_ more sense, thank you! 🐑

sjb9774 commented on src/cli/CliRandomizer.java:
```diff
@@ -0,0 +1,194 @@
+        List<String> allowedFlags = Arrays.asList("-i", "-o", "-s", "-d");
+        for (int i = 0; i < args.length; i++) {
+            if (allowedFlags.contains(args[i])) {
+                int flagValueIndex = i + 1;
```
Whoops, had removed in reverted commit, will get it back out of there

sjb9774 commented:
@Ajarmar Thanks for your attention and patience with this! I believe I've addressed all your comments but let me know if there's anything else needed.

Ajarmar commented on src/com/dabomstew/pkrandom/cli/CliRandomizer.java:
```diff
@@ -0,0 +1,188 @@
+
+
+                    Randomizer randomizer = new Randomizer(settings, romHandler, saveAsDirectory);
+                    randomizer.randomize(destinationRomFilePath);
```
My bad, the change I suggested for `fixFileName` doesn't quite make everything work. In order for the file name fix to have any effect, `randomize` has to use the path stored in the file handler returned by `fixFileName`. (The same thing is done in `NewRandomizerGUI`, [see here.](https://github.com/Ajarmar/universal-pokemon-randomizer-zx/blob/master/src/com/dabomstew/pkrandom/newgui/NewRandomizerGUI.java#L774)) With that in mind, `File fh = new File(destinationRomFilePath);` should be put before the `if (!saveAsDirectory)` clause.
```suggestion
                    randomizer.randomize(fh.getAbsolutePath());
```

Ajarmar commented on src/com/dabomstew/pkrandom/cli/CliRandomizer.java:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    private static void printUsage() {
+        System.err.println("Usage: java -jar PokeRandoZX.jar cli -s <path to settings file> -i <path to source ROM> -o <path for new ROM> [-d save as directory]");
```
You changed this one to print on `stderr`, but I think you missed the other change I suggested here.
```suggestion
        System.err.println("Usage: java -jar PokeRandoZX.jar cli -s <path to settings file> -i <path to source ROM> -o <path for new ROM> [-d]");
        System.err.println("-d: Save 3DS game as directory (LayeredFS)");
```

Ajarmar commented on src/com/dabomstew/pkrandom/newgui/NewRandomizerGUI.java:
```diff
@@ -25,6 +25,7 @@
 /*--  along with this program. If not, see <http://www.gnu.org/licenses/>.  --*/
 /*----------------------------------------------------------------------------*/
 
+import cli.CliRandomizer;
```
Need to change this import since the package was changed.
```suggestion
import com.dabomstew.pkrandom.cli.CliRandomizer;
```

Ajarmar reviewed (changes requested):
We're getting there, just a few more things that I found.

sjb9774 commented:
@Ajarmar Thanks again for your time. I've pushed up changes which should address your comments! I've also taken the liberty of adding support for a `--help` flag; since I had to step away from this for a few days I found myself needing a refresher on how the program worked and realized it didn't have an intuitive way to get the usage. There may be a world where getting the usage from `--help` specifically should output to `stdout` instead of `err` since it's expected behavior, just let me know if you think that's more appropriate here.

Ajarmar reviewed (approved):
This looks good as it is for now. I'll merge this in now, thanks!
